src/ducks/brandDictionary/withBrands.jsx

```diff
@@ -1,10 +1,23 @@
-import { connect } from 'react-redux'
+import { useMemo } from 'react'
+import { connect, useStore } from 'react-redux'
 
 import { getInstalledBrandsFromQuery } from './selectors'
 
-const withBrands = (options = { queryName: 'triggers' }) =>
-  connect(state => ({
+const mapStateToProps =
+  (options = { queryName: 'triggers' }) =>
+  state => ({
     brands: getInstalledBrandsFromQuery(options.queryName)(state)
-  }))
+  })
+
+const withBrands = (options = { queryName: 'triggers' }) =>
+  connect(mapStateToProps(options.queryName))
+
+export const useBrands = (options = { queryName: 'triggers' }) => {
```

Comment:
the refactor time, is also the time to test the component. I would test using `renderHook` to validate we keep the same behaviour 🙏

Reply:
ça peut être intéressant de vérifier que `useBrands` fonctionne comme `withBrands`, par contre je vois pas trop comment, as-tu une idée ? 🤔

Reply:
personnellement, je fais confiance dans cozy-client pour faire le taf. Et donc je le mocke dans la plupart des tests.

Mais lorsque je souhaite vérifier des fonctionnements (ce qui me semble être ton cas), je m'attelle à une use case minimum). Quelle est la donnée que tu t'attends à recevoir exactement de useBrands ?

Reply:
Ben justement c'est là ou je vois pas trop quoi tester, puisqu'au final on va se retrouver à mocker la réponse, non ? Je m'attends à recevoir un tableau d'objet.

Reply:
import { renderHook } from '@testing-library/react-hooks'
import { useBrands } from './withBrands'
import { getInstalledBrandsFromQuery } from './selectors'

jest.mock('./selectors', () => ({
  getInstalledBrandsFromQuery: jest.fn().mockReturnValue(() => 'brands')
}))

jest.mock('react-redux', () => ({
  useStore: jest.fn(),
  createSelectorHook: jest.fn()
}))

describe('useBrands', () => {
  it('should return getInstalledBrandsFromQuery from state', () => {
    // When
    const render = renderHook(useBrands)

    // Then
    expect(render.result.current).toEqual({ brands: 'brands' })
  })

  it('should call getInstalledBrandsFromQuery with undefined', () => {
    // When
    renderHook(useBrands)

    // Then
    expect(getInstalledBrandsFromQuery).toHaveBeenCalledWith(undefined)
    expect(getInstalledBrandsFromQuery).toHaveBeenCalledTimes(1)
  })

  it('should call getInstalledBrandsFromQuery when queryName is defined', () => {
    // When
    renderHook(() => useBrands({ queryName: 'test' }))

    // Then
    expect(getInstalledBrandsFromQuery).toHaveBeenCalledWith(undefined)
    expect(getInstalledBrandsFromQuery).toHaveBeenCalledTimes(1)
  })
})

Reply:
Par contre, suite à ces tests, je ne comprends pas :
- pourquoi getInstalledBrandsFromQuery est appelé avec undefined. 
- pourquoi on utilise useMemo() (sûrement lié au point précédent)
- pourquoi dans withBrands.jsx ligne 7, on définit une default option {queryName: 'triggers'} qui est déjà définie par les 2 fonctions qui appellent cette méthode privée.

Ajouter des tests permet de répondre à ces interrogations, et fixer les soucis. 🙏

Reply:
- j'ai pas saisi, tu parles du dernier `expect(getInstalledBrandsFromQuery).toHaveBeenCalledWith(undefined)` dans ton dernier test ?
- je comprends pas l'interrogation sur le useMemo 🤔 il te semble inutile ? Pourquoi ?
- j'ai refacto un `mapStateToProps = options => state` devrait suffire, c'est ce que tu avais en tête ?

Reply:
d'après ce que je lis : 
`getInstalledBrandsFromQuery(options.queryName)(state)` 

la fonction getInstalledBrandsFromQuery n'est pas appelée avec undefined.

Alors pourquoi le test reçoit undefined. (les console log montrent aussi que c'est undefined qui arrivent à cet endroit)

Reply:
> je comprends pas l'interrogation sur le useMemo 🤔 il te semble inutile ? Pourquoi ?

Pourquoi y aurait-il besoin d'un useMemo ?
Je ne vois aucune problématique de perf' à vue de nez.

Reply:
- useMemo : c'est pas forcément une problématique de perf, c'est aussi pour éviter un rerender si le hook renvoie le même résultat, non ? Je pense que c'est l'idée sur le `withFilters` dont j'ai repompé l'approche

Reply:
Vous pouvez vous appuyer sur useMemo comme un moyen d’optimiser les performances, mais pas comme une garantie sémantique. À l’avenir, React pourrait choisir « d’oublier » certaines valeurs précédemment mémoïsées et de les recalculer au rendu suivant, par exemple pour libérer la mémoire exploitée par des composants présents hors de l’écran. Écrivez votre code de façon à ce qu’il fonctionne sans useMemo et ajoutez-le ensuite pour optimiser les performances.

d'après : 
https://fr.reactjs.org/docs/hooks-reference.html#usememo

Je pense qu'on n'a pas besoin de useMemo.

Reply:
super pour le undefined. C'était un bug : on passait options.queryName au lieu d'options, c'est ça ?

Reply:
- useMemo : ok bon je comprends pas trop mais enlevons-le pour avancer, un PdC serait le bienvenu.
- undefined : l'arg n'était pas `options` mais `queryName` en fait